org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/SaveActionExtensions.scala

```diff
@@ -102,18 +120,39 @@ trait SaveActionExtensions extends HasLogger {
    * Applies all save actions to the contents of the given document.
    */
   private def applySaveActions(udoc: IDocument): Unit = {
-    val undoManager = DocumentUndoManagerRegistry.getDocumentUndoManager(udoc)
-    undoManager.beginCompoundChange()
+    def updateEditor() = EditorUtils.doWithCurrentEditor {
+      _.selectAndReveal(lastSelection.getOffset, lastSelection.getLength)
+    }
 
-    applyDocumentExtensions(udoc)
+    EditorUtils.withScalaSourceFileAndSelection { (ssf, sel) =>
+      lastSourceFile = ssf
+      lastSelection = sel
 
-    undoManager.endCompoundChange()
+      val undoManager = DocumentUndoManagerRegistry.getDocumentUndoManager(udoc)
+      undoManager.beginCompoundChange()
+
+      try applyDocumentExtensions(udoc)
+      catch {
+        // thrown when save actions had to be aborted
+        case e: IllegalStateException =>
+          updateEditor()
+          throw e
```

Comment:
Is there no need to call `endCompoundChange()` when an exception is thrown? Or should we call it inside a `finally` block?